README.md
# ansible-role-test

`ansible-role-test` is a tool that uses docker containers to quickly create
disposable boxes that can be provisioned by a test ansible playbook via ssh.

## Requirements

* docker
* python 2.7+, 3.x

## Installation

* Clone repository
* Optional: Activate your virtualenv or run `make venv` to create a local one
* Run `make install`
* Either run `ansible-role-test` or `venv/bin/ansible-role-test`

## Development

Instead of running `make install`, run `make dev` and use `venv/bin/ansible-role-test`.

## Usage

```
Usage: ansible-role-test [OPTIONS] COMMAND [ARGS]...

  ansible-role-test is a docker based testing utility for ansible roles.

Options:
  --help  Show this message and exit.

Commands:
  init       Create the test folder with a default test...
  snapshots  Manipulate saved containers
  test       Run tests ROLE can be either be a local path,...
```

### `init` command

Creates a `tests` folder with a default `main.yml` file that looks like
the one shown in the [Writing tests] section below.

```
Usage: ansible-role-test init [OPTIONS] [PATH]

  Create the test folder with a default test file

Options:
  -h, --help  Show this message and exit.
```

### `test` command

```
Usage: ansible-role-test test [OPTIONS] ROLE

  Run tests

  ROLE can be either be a local path, a git repository or an ansible-galaxy
  role name.

Options:
  -c, --config FILENAME           Config file to use for the tests
  --roles-path ROLES_PATH         Search path for non-galaxy roles that might
                                  be required as dependencies
  --library-path LIBRARY_PATH     Search path for custom ansible modules
  --plugins-action-path PLUGINS_ACTION_PATH
                                  Search path for custom action plugins
  --plugins-filter-path PLUGINS_FILTER_PATH
                                  Search path for custom filter plugins
  --plugins-lookup-path PLUGINS_LOOKUP_PATH
                                  Search path for custom lookup plugins
  -e, --extra-vars EXTRA_VARS     Set additional variables as key=value or
                                  YAML/JSON
  -l, --limit SUBSET              Limit selected hosts to a given pattern
  --skip-tags SKIP_TAGS           Only run plays and tasks whose tags do not
                                  match these values
  -t, --tags TAGS                 Only run plays and tasks tagged with these
                                  values
  -v                              Verbose mode (-vvv for more, -vvvv to enable
                                  connection debugging)
  --ansible-version ANSIBLE_VERSION
                                  The ansible version to use (either 1.8, 1.9
                                  or latest)
  --privileged                    Run test containers in privileged mode
                                  (dangerous)
  --cache                         Cache yum/apt folders on the host
  --save [failed|successful|all]  Save containers, can be either one of
                                  "failed", "successful" and "all"
  -h, --help                      Show this message and exit.
```

```bash
# Create a default test file and run it on a local role
ansible-role-test init /path/to/role
ansible-role-test test --save=failed /path/to/role
# Test a git repository
ansible-role-test test https://github.com/org/repo.git
# Test a branch on a git repository
ansible-role-test test https://github.com/org/repo.git#my-branch
# Test an ansible-galaxy role
ansible-role-test test user.role
```

### `snapshots` command

Helper commands to inspect saved images generated by the `test` command.

```
Usage: ansible-role-test snapshots [OPTIONS] COMMAND [ARGS]...

  Manipulate saved containers

Options:
  -h, --help  Show this message and exit.

Commands:
  list   List all snapshots and whether they failed or...
  purge  Delete all snapshots
  rm     Remove a specific snapshot
  view   Display the output of the ansible play that...
```

#### `snapshots list` command

List all snapshots currently saved in the local repository.

Example output:

```
ROLE NAME               CONTAINER           STATUS          DATE                    IMAGE NAME
aeriscloud.nodejs       centos-6            successful      2015-06-10T12:58:52     aeriscloud.nodejs.centos-6:successful-1433908732
aeriscloud.nodejs       centos-7            failed          2015-06-10T12:58:47     aeriscloud.nodejs.centos-7:failed-1433908727
```

#### `snapshots view` command

Display the output of the ansible job that was run on this host.

#### `snapshots rm` command

Removes a stored snapshot by name, eg `ansible-role-test snapshots rm aeriscloud.nodejs.centos-6:successful-1433908732`

#### `snapshots purge` command

Deletes all stored snapshots.

## Writing tests

Create a `tests` folder in your role, then in that folder create a yaml file that
looks like that one:

```yaml
---
# The name of your test
name: "My test name"
# The containers you will use for this test, you could for example spawn a debian
# centos, ubuntu, etc... and run the tests on "all" hosts
# If skipped, tests will be run on every containers available
#containers:
#  master1: 'centos:6'
#  slave1:
#    image: 'centos:7'
#    vars:
#      host_var1: foobar # defines host_var1 on this host on particular
#  slave2: 'centos:7'
#  slave3: 'debian:wheezy'
# You can also setup custom inventory groups to be declared in the inventory, if
# no containers are declared above, they will be merged with the default groups.
# For exemple the following declaration will create 2 groups, master and slaves.
#groups:
#  masters:
#  - master1
#  slaves:
#  - slave1
#  - slave2
#  - slave3
# This is your test playbook
playbook:
- hosts: all
# You should have your role called in roles, with the proper variables set, you
# can call the same role several times in a row but I'd rather recommend creating
# separate test files for each call to ensure that they run in a clean env
  roles:
# "@ROLE_NAME@" is a magic variable that will be replaced by the name of the role
# on your filesystem before running the tests
  - role: "@ROLE_NAME@"
    var1: something
    var2: something
# You should verify that your role executed properly here, using tasks
  tasks:
  - name: "Check that my-role did that thing properly"
    module: do-something
```

## Saving containers

If desired, it is possible to use the `--save` option to save containers at the
end of each test. Possible values for this option are `failed`, `successful` and
`all` and will filters which containers to save. One useful use case is to save
`failed` containers to inspect what file/package could be missing.

Containers are saved as `art/{ROLE NAME}.{CONTAINER NAME}` with the tag
`{STATUS}-{STAMP}`.

Saved images can be found by running `ansible-role-test snapshots list`.
It is possible to inspect the image by running:

```bash
$ docker run -t -i art/role-name.system:tag bash
[root@671165bdfcb0 /]#
```

Once done, one can removed failed containers by running `docker rmi art/foo:tag`
or `ansible-role-test snapshots rm art/foo:tag`. A faster way to remove all
the corresponding images is to run `ansible-role-test snapshots purge`.

## Paths and config file

Most of the time, your roles might depend on other local roles or plugins, in
which case you can use the `--role-path`, `--library-path`, etc... flags to help
ansible find those. In some cases though it might result in unwieldy commands
due to the sheer amount of flags passed around. When that happens you can use a
config file that follows this format to provide the proper information to ansible:

```yaml
---
roles: ansible/roles
library: ansible/library
plugins:
  action: ansible/plugins/actions
  filter: ansible/plugins/filters
```

Then call `ansible-role-test` with the `--config` flag pointing to this file.
The given paths are relative to the config file's location.

## Available test containers

You can find them on the wizcorp user on the docker registry, they should be
named `ansible-<distro>:<version>`.

`image name`    | `name in playbook`
----------------|-------------------
centos:6        | centos-6
centos:7        | centos-7
debian:wheezy   | debian-wheezy
debian:jessie   | debian-jessie
ubuntu:14.04    | ubuntu-lts
ubuntu:15.04    | ubuntu-15

They are also automatically separated in 3 groups `centos`, `debian` and `ubuntu`.

Please check [`aeriscloud@docker`](https://registry.hub.docker.com/repos/aeriscloud/)
for the current list of available images.

If you wish to build all the images locally, you can run `make docker`.

## Known issues/caveats

* Any role that deals with kernel modules or raw hardware (eg. `iptables` or
  formatting disks) will run into issues while running in docker.
  `iptables` seems to work in privileged mode with no side effects but
  anything that does destructive operations should not be tested with
  privileged mode on.
* While the centos boxes are running `systemd-container`, debian and ubuntu
  boxes are not and as such calls to systemctl/initd might not behave the
  same way as a real system.

## Disclaimer

This is currently a work in progress, I am not responsible and shall not
be held responsible in any manner if this tool causes loss of data, hardware
faults, act of gods, invocation of old or ancient ones, elder gods and other
horrors from the depths.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.
This project adheres to [Semantic Versioning](http://semver.org/).

## 0.2.0 - [unreleased]
### Added
* Containers can now use custom images outside of the aeriscloud namespace

### Fixed
* `fr` instead or `fg` being used in some `click.secho` calls
* Properly show progress when downloading ansible image
* `systemd` for Centos 7 now properly works
* Unicode issue when reading progress from docker on Python 3.x

### Removed
* `--cache` flag was removed due to it acting weird and the performance gain
  being nearly non-existent

## [0.1.0] - 2015-06-16
### Added
* Volumes declared on test containers can now be cached between calls by using
  the `--cache` flag (cached in your user's cache folder)
* Support for creating inventory groups in the test file
* Provisionned containers can now be committed to the local registry by using
  the `--save` option, the option takes either `failed`, `successful` or `all`
  to filter which containers to save
* Can add host variables by declaring a container as a map with the `image`
  and `vars` key
* `snapshots` subcommand for listing, inspecting and deleting saved containers

### Changed
* `centos` images now have some extra packages installed to make them closer
  to a normal instance
* Docker version is detected by the API at start
* Dropped `make dist` and `pyinstaller` due to the build breaking on a semi
  regular basis
* Renamed `make build` to `make install`

### Fixed
* Prevent `make docker` from picking up `docker/Makefile` as a target.
* Bug when trying to load an ansible-galaxy role with `--roles-path` not set (again)
* Properly count failed and successful tests
* Crash when `Dead` is not returned by the inspect API
* `ansible-galaxy` crash when a dependency was installed twice

## [0.0.5] - 2015-05-22
### Changed
* Flags now override the config, so it is possible to so things such as
  `-c config.yml --roles-path=/path/to/roles` and have the --roles-path
  override the config

### Fixed
* A bug when not all paths are specified when a config file is used

## [0.0.4] - 2015-05-22
### Added
* Pull images that are not available locally during tests
* New docker-pull target in the Makefile
* New `--config flag` to provide preconfigured role, library and plugin folders

### Changed
* Can customize the virtualenv binary in the Makefile
* Bindings on the ansible box are now read-only
* Merge all docker Makefiles in a single one

### Fixed
* Use a temporary dir in the user path (boot2docker support)
* Annoying log output on osx about urllib

## [0.0.3] - 2015-05-20
### Added
* Added changelog
* Python 3.x support using six

### Changed
* Cleaned up Makefile
* Added some extra data in setup.py

### Fixed
* Fixed duplicate @click.command in cli.py

## [0.0.2] - 2015-05-18
### Added
* Added docker images on docker registry
* Added `--privileged` option

### Changed
* Cleaned up README
* Better option documentation
* Containers are not started in privileged mode anymore
* Changed httpredir mirror for debian to cloudfront mirror

### Fixed
* Bug when trying to load an ansible-galaxy role with `--roles-path` not set
* Added dbus to centos:7 image
* Fixed resolvconf package on debian

## [0.0.1] - 2015-05-15
### Added
* Initial version

[unreleased]: https://github.com/AerisCloud/ansible-role-test/compare/v0.1.0...HEAD
[0.1.0]: https://github.com/AerisCloud/ansible-role-test/compare/v0.0.5...v0.1.0
[0.0.5]: https://github.com/AerisCloud/ansible-role-test/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/AerisCloud/ansible-role-test/compare/v0.0.3...v0.0.4
[0.0.3]: https://github.com/AerisCloud/ansible-role-test/compare/v0.0.2...v0.0.3
[0.0.2]: https://github.com/AerisCloud/ansible-role-test/compare/v0.0.1...v0.0.2
[0.0.1]: https://github.com/AerisCloud/ansible-role-test/tree/v0.0.1


ansibleroletest/cli.py
from __future__ import print_function

import click
import logging
import os
import six
import sys
import yaml

from .container import ContainerManager
from .docker import client as docker_client
from .framework import TestFramework

logging.captureWarnings(True)


@click.command(context_settings={'help_option_names': ['-h', '--help']})
# path options
@click.option('-c', '--config', default=None,
              help='Config file to use for the tests',
              type=click.File(mode='rb'))
@click.option('--roles-path', default=None,
              metavar='ROLES_PATH',
              help='Search path for non-galaxy roles that might be required '
                   'as dependencies')
@click.option('--library-path', default=None,
              metavar='LIBRARY_PATH',
              help='Search path for custom ansible modules',
              envvar='ANSIBLE_LIBRARY')
@click.option('--plugins-action-path', default=None,
              metavar='PLUGINS_ACTION_PATH',
              help='Search path for custom action plugins',
              envvar='ANSIBLE_ACTION_PLUGINS')
@click.option('--plugins-filter-path', default=None,
              metavar='PLUGINS_FILTER_PATH',
              help='Search path for custom filter plugins',
              envvar='ANSIBLE_FILTER_PLUGINS')
@click.option('--plugins-lookup-path', default=None,
              metavar='PLUGINS_LOOKUP_PATH',
              help='Search path for custom lookup plugins',
              envvar='ANSIBLE_LOOKUP_PLUGINS')
# ansible options
@click.option('-e', '--extra-vars', multiple=True,
              metavar='EXTRA_VARS',
              help='Set additional variables as key=value or YAML/JSON')
@click.option('-l', '--limit',
              metavar='SUBSET',
              help='Limit selected hosts to a given pattern')
@click.option('--skip-tags', default=None,
              metavar='SKIP_TAGS',
              help='Only run plays and tasks whose tags do not match these '
                   'values')
@click.option('-t', '--tags', default=None,
              metavar='TAGS',
              help='Only run plays and tasks tagged with these values')
@click.option('-v', 'verbosity', count=True,
              help='Verbose mode (-vvv for more, -vvvv to enable connection '
                   'debugging)')
# extra
@click.option('--ansible-version', default='latest',
              metavar='ANSIBLE_VERSION',
              help='The ansible version to use (either 1.8, 1.9 or latest)',
              type=click.Choice(['1.8', '1.9', 'latest']))
@click.option('--privileged', is_flag=True, default=False,
              help='Run test containers in privileged mode (dangerous)')
@click.option('--cache', is_flag=True,
              help='Cache yum/apt folders on the host')
@click.option('--save-failed/--no-save-failed', default=False,
              help='Save failed containers for inspection')
@click.argument('role')
def main(role,
         config,
         # path args
         roles_path, library_path, plugins_action_path,
         plugins_filter_path, plugins_lookup_path,
         # ansible-playbook args
         extra_vars, limit, skip_tags, tags, verbosity,
         # misc
         ansible_version, privileged, cache, save_failed):
    """
    ansible-role-test is a docker based testing utility for ansible roles.

    ROLE can be either be a local path, a git repository or an ansible-galaxy
    role name.
    """
    with ContainerManager(docker_client()) as docker:
        ansible_paths = {
            'roles': roles_path,
            'library': library_path,
            'plugins': {
                'action': plugins_action_path,
                'filter': plugins_filter_path,
                'lookup': plugins_lookup_path,
            }
        }

        _load_config(ansible_paths, config)

        framework = TestFramework(docker, role, ansible_paths,
                                  ansible_version)
        res = framework.run(
            extra_vars=extra_vars,
            limit=limit,
            skip_tags=skip_tags,
            tags=tags,
            verbosity=verbosity,
            privileged=privileged,
            cache=cache,
            save_failed=save_failed
        )

        if res != 0 and not save_failed:
            click.secho('''
info: some of the tests have failed. If you wish to inspect the failed
      containers, rerun the command while adding the --save-failed flag
      to your command line.''', fg='blue')
    sys.exit(res)


def _load_config(conf, config_file=None):
    if not config_file:
        return

    base = os.path.dirname(config_file.name)
    content = yaml.load(config_file)

    # merge both objects if the original value is none
    def _update(obj_from, obj_to):
        for k, v in six.iteritems(obj_to):
            if isinstance(v, dict):
                _update(obj_from[k], v)
            elif v is None and k in obj_from and obj_from[k]:
                obj_to[k] = os.path.join(base, obj_from[k])

    _update(content, conf)

if __name__ == '__main__':
    main()
